File: resources/manualHttpRequests.js
require('dotenv').config()
const axios = require("axios");
const fs = require('fs');

const publicKey = process.env.OA_PUBLIC_KEY;
const secretKey = process.env.OA_SECRET_KEY;
const ACCESS_TOKEN_URL = process.env.ACCESS_TOKEN_URL
const AGENDA_UID = process.env.AGENDA_UID
const tokenFilePath = 'secret_token.json';
// const TBD_LOCATION_UID = process.env.TBD_LOCATION_UID

const retrieveAccessToken = async (apiSecretKey) => {

    // check if token already exist
    if (fs.existsSync(tokenFilePath)) {
        const tokenFileContent = await fs.promises.readFile(tokenFilePath, 'utf8');
        const tokenData = JSON.parse(tokenFileContent);

        if (tokenData.access_token && tokenData.endate && Date.now() - tokenData.endate < 3600000) {
                // console.log("Using existing access token");
            const accessToken = tokenData.access_token;
            return accessToken;
        }
    };
        
    //else 
    console.log("Request a new token and save it in secret_token.json");
    const headers = {
        "Content-Type": 'application/json',
    };
    const body = {
        grant_type: "client_credentials",
        code: apiSecretKey,
    };

    try {
        const oauthResponse = await axios.post(ACCESS_TOKEN_URL, body, { headers });

        if (oauthResponse.status >= 200 && oauthResponse.status <= 299) {
            // Save token to token.json file
            const tokenData = {
                access_token: oauthResponse.data.access_token,
                endate: Date.now() + 3600000, // 1 hour from now
            };
            await fs.promises.writeFile(tokenFilePath, JSON.stringify(tokenData));

            return oauthResponse.data.access_token;
        }
    } catch (exc) {
        console.error("Error retrieving access token: ", exc.response ? exc.response.data : exc.message);
    }

    console.log("[Retrieve accessToken] Error - no access token retrieved");
    return null;
}

const getLocations = async (accessToken) => {

    const url = `https://api.openagenda.com/v2/agendas/${AGENDA_UID}/locations`;

    try {
        let after=0
        let all_locations=[]
        
        const headers = {
            "Content-Type": 'application/json',
            "access-token": accessToken,
            "nonce": Date.now(),
            };
        const params = {
            "after":after,
            "detailed":1
        };

        while (after != null ){
            headers["nonce"] = Date.now();
            params["after"]  = after
            const response   = await axios.get(url, { headers, params });
                if (response.status >= 200 && response.status <= 299) {
                    all_locations =all_locations.concat(response.data.locations);
                    after=response.data.after
                }
            
        }
        return all_locations

    } catch (exc) {
        console.error("Error retrieving locations: ", exc.response ? exc.response.data : exc.message);
    }
}

const postLocation = async (accessToken, address) => {
    const headers = {
        "Content-Type": 'application/json',
        "access-token": accessToken,
        "nonce": Date.now(),
    };
    const body = {
        name: address,
        address: address,
        countryCode: "FR",
        state: 0, // means "not verified"
    };
    const url = `https://api.openagenda.com/v2/agendas/${AGENDA_UID}/locations`;

    const response = await axios.post(url, body, { headers })
        .catch(function (error) {
            if (error.response) {
                // Catching OA response
                if (error.response.status = 400 
                    && error.response.data.message == "geocoder didn't find address") {
                    console.error("No existing address found by OA API");
                    return null;
                }
                console.log(error.response.status, error.response.data);
            } else {
                // Something happened in setting up the request that triggered an Error
                console.log('Error OA Post location', error);
            }

        });
        return response?.data ? response.data : null
}

const deleteLocation = async (accessToken, locationUid) => {
    const headers = {
        "Content-Type": 'application/json',
        "access-token": accessToken,
        "nonce": Date.now(),
    };

    const url = `https://api.openagenda.com/v2/agendas/${AGENDA_UID}/locations/${locationUid}`;

    try {
        const response = await axios.delete(url, { headers });
        if (response.status >= 200 && response.status <= 299) {
            // console.log("Locations Response: ", response.data);
            return response.data;
        }
    } catch (exc) {
        console.error("Error deleting location: ", exc.response ? exc.response.data : exc.message);
    }
}

const createEvent = async (accessToken, event) => {
    const headers = {
        "Content-Type": 'application/json',
    };
    const body = {
        "access_token": accessToken,
        "nonce": Date.now(),
        "data": event,
    };
    const url = `https://api.openagenda.com/v2/agendas/${AGENDA_UID}/events`;
    try {
        const eventCreationResponse = await axios.post(url, body, { headers });

        if (eventCreationResponse.status >= 200 && eventCreationResponse.status <= 299) {
            console.log("Event creation response: ", eventCreationResponse.data);
            return eventCreationResponse.data;
        }
    } catch (exc) {
        console.error("Error retrieving access token: ", exc.response ? exc.response.data : exc.message);
    }
}

const deleteEvent = async (accessToken, eventUid) => {
    const headers = {
        "Content-Type": 'application/json',
        "access-token": accessToken,
        "nonce": Date.now(),
    };
    const url = `https://api.openagenda.com/v2/agendas/${AGENDA_UID}/events/` + eventUid.toString();
    try {
        const response = await axios.delete(url, { headers });
        if (response.status >= 200 && response.status <= 299) {
            console.log("Delete Event Response: ", response.data);
            return response.data;
        }
    } catch (exc) {
        console.error("Error deleting event: ", exc.response ? exc.response.data : exc.message);
    }
}

module.exports = { retrieveAccessToken, getLocations, postLocation, createEvent, deleteEvent, deleteLocation }

// (async () => {
//     console.log("START");

//     const accessToken = await retrieveAccessToken(secretKey);
//     if (accessToken) {
//         console.log("Access Token: ", accessToken);

//         // const locations = await getLocations(accessToken);
//         // console.log("Locations: ", locations);
//         // const event = {
//         //     title: {fr: "[TEST] Evénement ACN"},
//         //     description: {fr: "Description courte ACN"},
//         //     timings: [
//         //         {
//         //             begin: "2024-06-18T17:00:00+0200",
//         //             end: "2024-06-18T19:00:00+0200"
//         //         }
//         //     ],
//         //     locationUid: TBD_LOCATION_UID,
//         //     attendanceMode: 1
//         // }
//         // await createEvent(accessToken, event)
//         await deleteEvent(accessToken, 38636380)
//     }

//     console.log("END");
// })();

File: resources/python/manualHttpRequests.py
"""_summary_
Function to interact with OpenAgenda API
"""

import math, random
import os
import requests, json
import time

# Chargement des variables d'environnement
public_key = os.getenv('OA_PUBLIC_KEY')
secret_key = os.getenv('OA_SECRET_KEY')
ACCESS_TOKEN_URL = os.getenv('ACCESS_TOKEN_URL')
AGENDA_UID = os.getenv('AGENDA_UID')
token_file_path = './secret_token.json'
TBD_LOCATION_UID = os.getenv('TBD_LOCATION_UID')


def get_nonce():
    """
    A timestamp + random number to be unique for each request, even those in short time interval (with same timestamp)
    """
    nonce = str(int(time.time())) + str(math.floor(random.random()*1000))
    return nonce
    

def retrieve_access_token(api_secret_key):
    # Vérifier si le jeton existe déjà
    if os.path.exists(token_file_path):
        with open(token_file_path, 'r', encoding='utf8') as token_file:
            token_data = json.load(token_file)
        time_diff =token_data['endate']- time.time()
        if ('access_token' in token_data) and ('endate' in token_data) and (time_diff > 1800):
            return token_data['access_token']
        
    print("Request a new token and save it in secret_token.json")
    headers = {
        "Content-Type": 'application/json',
    }
    body = {
        "grant_type": "client_credentials",
        "code": api_secret_key,
    }

    try:
        oauth_response = requests.post(ACCESS_TOKEN_URL, json=body, headers=headers)
        oauth_response.raise_for_status()

        token_data = {
            'access_token': oauth_response.json()['access_token'],
            'endate': int(time.time()) + oauth_response.json()['expires_in'] - 600,  # 50m à partir de maintenant
        }

        with open(token_file_path, 'w', encoding='utf8') as token_file:
            json.dump(token_data, token_file)

        return token_data['access_token']

    except requests.exceptions.RequestException as exc:
        print(f"Error retrieving access token: {exc}")
        return None

def get_locations(access_token):

    url = f"https://api.openagenda.com/v2/agendas/{AGENDA_UID}/locations"
    after =0
    all_locations=[]
    while after is not None:
        try:
            headers = {
                    "Content-Type": 'application/json',
                    "access-token": access_token,
                    "nonce": get_nonce()
                    }
            response = requests.get(url, headers=headers, params={'after': after})
            response.raise_for_status()
            locations_part=json.loads(response.text)
            all_locations.extend(locations_part.get('locations'))
            after=locations_part.get('after')

        except requests.exceptions.RequestException as exc:
            print(f"Error retrieving locations: {exc}")
            return None
    return all_locations

def post_location(access_token, name, adresse):
    """
    Create a new location in OpenAgenda, using OA Geocoder with 'name' and 'address' as search parameters
    Args:
        access_token (str): The access token obtained by calling `retrieve_access_token`
        name (str): The name of the new location
        adresse (str): The address of the new location
    Returns:
        The JSON response of the API call
    """
    headers = {
        "Content-Type": 'application/json',
        "access-token": access_token,
        "nonce": get_nonce(),
    }
    body = {
        "name": name,
        "address": adresse,
        "countryCode": "FR",
        "state": 0,  # signifie "non vérifié"
    }
    url = f"https://api.openagenda.com/v2/agendas/{AGENDA_UID}/locations"

    try:
        response = requests.post(url, json=body, headers=headers)
        response.raise_for_status()
        return response.json()

    except requests.exceptions.RequestException as exc:
        text_json = json.loads(exc.response.text)
        if text_json.get('message') == "geocoder didn't find address":
            print("No existing address found by OA API")
        else:
            print(f"Error Posting location on OA: {exc}")
        return None

def delete_location(access_token, location_uid):
    headers = {
        "Content-Type": 'application/json',
        "access-token": access_token,
        "nonce": get_nonce(),
    }
    url = f"https://api.openagenda.com/v2/agendas/{AGENDA_UID}/locations/{location_uid}"

    try:
        response = requests.delete(url, headers=headers)
        response.raise_for_status()
        return response.json()

    except requests.exceptions.RequestException as exc:
        print(f"Error deleting location: {exc}")
        return None

def get_events(public_key, params: dict):
    headers = {
        "Content-Type": 'application/json',
        "nonce": get_nonce()
    }
    url = f"https://api.openagenda.com/v2/agendas/{AGENDA_UID}/events?key={public_key}"
    try:
        response = requests.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    except requests.exceptions.RequestException as exc:
        print(f"Error getting events: {exc}")
        return None

def create_event(access_token, event, image_path=None):
    headers = {
        "Content-Type": "application/json",
        "access-token": access_token,
        "nonce": get_nonce(),
    }
    body = event
    url = f"https://api.openagenda.com/v2/agendas/{AGENDA_UID}/events"
    # files = None
    # if image_path:
    #     with open(image_path, 'rb') as img_file:
    #         files = {'image': img_file}
    try:
        event_creation_response = requests.post(url, json=body , headers=headers)
        
        if event_creation_response.status_code != 200:
            print(f"Error creating event: Status Code {event_creation_response.status_code}")
            print(f"Response:")
            print(json.dumps(event_creation_response.json(), indent=4))
            return None
        print('event "'+ event['title']['fr'] + '" created with uid: ' + str(json.loads(event_creation_response.text)['event']['uid']) )
        return  event_creation_response.json()

    except requests.exceptions.RequestException as exc:
        print(f"Error creating event: {exc}")
        return None


def delete_event(access_token, event_uid):
    headers = {
        "Content-Type": 'application/json',
        "access-token": access_token,
        "nonce": get_nonce(),
    }
    url = f"https://api.openagenda.com/v2/agendas/{AGENDA_UID}/events/{event_uid}"

    try:
        response = requests.delete(url, headers=headers)
        response.raise_for_status()
        return response.json()

    except requests.exceptions.RequestException as exc:
        print(f"Error deleting event: {exc}")
        return None
